apps/src/lib/ui/SyncOmniAuthSectionControl.jsx

```diff
@@ -0,0 +1,151 @@
+import React, {PropTypes} from 'react';
+import {connect} from 'react-redux';
+import i18n from '@cdo/locale';
+import {OAuthSectionTypes} from '../../templates/teacherDashboard/shapes';
+import {
+  importRoster,
+  sectionCode,
+  sectionProvider
+} from '../../templates/teacherDashboard/teacherSectionsRedux';
+import Button, {ButtonColor, ButtonSize} from '../../templates/Button';
+
+const PROVIDER_NAME = {
+  [OAuthSectionTypes.clever]: i18n.loginTypeClever(),
+  [OAuthSectionTypes.google_classroom]: i18n.loginTypeGoogleClassroom(),
+};
+
+export const READY = 'ready';
+export const IN_PROGRESS = 'in-progress';
+export const SUCCESS = 'success';
+export const FAILURE = 'failure';
+
+/**
+ * Button that will re-sync an omniauth section's roster with the third-paty
+ * provider. This component owns that logic (along with the redux module).
+ * It in turn renders a button component that handles display states.
+ */
+class SyncOmniAuthSectionControl extends React.Component {
+  static propTypes = {
+    sectionId: PropTypes.number.isRequired,
+    // Provided by Redux
+    sectionCode: PropTypes.string,
+    sectionProvider: PropTypes.oneOf(Object.values(OAuthSectionTypes)),
+    importRoster: PropTypes.func.isRequired,
+  };
+
+  state = {
+    buttonState: READY,
+  };
+
+  static reloadPage() {
+    // Extracted so we can stub it in tests
+    window.location.reload();
+  }
+
+  onClick = () => {
+    const {sectionCode, importRoster} = this.props;
+    const {buttonState} = this.state;
+    if ([IN_PROGRESS, SUCCESS].includes(buttonState)) {
+      // Don't acknowledge click events while request is in progress.
+      // For now, ignore them on success too - the page reload will take care of it.
+      return;
+    }
+
+    if (buttonState === FAILURE) {
+      // On click after failure, reset the button so the user can try again.
+      this.setState({buttonState: READY});
+      return;
+    }
+
+    // Default case: Button is READY
+    this.setState({buttonState: IN_PROGRESS});
+    // Section code is the course ID, without the G- or C- prefix.
+    const courseId = sectionCode.replace(/^[GC]-/, '');
+    importRoster(courseId)
```

Comment:
Should we name this differently to make that a little more explicit? `importOrUpdateRoster`? (or something more concise). If nothing else, we could just rename the prop in this particular component to be `updateRoster`